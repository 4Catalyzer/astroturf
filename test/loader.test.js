const { runLoader } = require('./helpers');

describe('webpack loader', () => {
  it('should add imports', async () => {
    const [code] = await runLoader(
      `
      import { css } from 'astroturf';

      const styles = css\`\`

      function Button() {
        return (
          <button
            css={css\`
              color: blue;
            \`}
          />
        );
      }

      export default Button;
    `,
      { enableCssProp: true },
    );

    expect(code).toContain('/** @jsx _j **/');
    expect(code).toContain('/** @jsxFrag _f **/');

    expect(code).toContain(
      'import _CssProp1_button from "./MyStyleFile-CssProp1_button.css"',
    );
    expect(code).toContain('[_CssProp1_button');

    expect(code).toContain('import _styles from "./MyStyleFile-styles.css"');
    expect(code).toContain('const styles = _styles;');
  });

  it('finds different css tag names', async () => {
    const [, styles] = await runLoader(
      `
      import { css as less } from 'astroturf';

      const styles = less\`
        .bar {
          font-weight: bold;
          font-style: italic;
        }
      \`;
    `,
      {
        tagName: 'less',
        extension: '.less',
      },
    );

    expect(styles).toHaveLength(1);
    expect(styles[0].absoluteFilePath.endsWith('.less')).toBe(true);
  });

  it('finds different styled tag names', async () => {
    const [, styles] = await runLoader(
      `
      const styles = s('div')\`
        font-weight: bold;
        font-style: italic;
      \`;
    `,
      {
        styledTag: 's',
      },
    );

    expect(styles).toHaveLength(1);
  });
});
