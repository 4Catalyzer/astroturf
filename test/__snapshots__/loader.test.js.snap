// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader css-no-assignment js : Compilation Error 1`] = `
"There are multiple anonymous css tags that would conflict. Differentiate each tag by assigning the output to a unique identifier
[0m [90m  7 | [39m[32m\`[39m[33m;[39m[0m
[0m [90m  8 | [39m[0m
[0m[31m[1m>[22m[39m[90m  9 | [39mcss[32m\`[39m[0m
[0m [90m    | [39m[31m[1m^[22m[39m[0m
[0m [90m 10 | [39m[32m  .red {[39m[0m
[0m [90m 11 | [39m[32m    color: red;[39m[0m
[0m [90m 12 | [39m[32m  }[39m[0m"
`;

exports[`Loader display-name 0: display-name-FooBarBaz.css 1`] = `
".foo-bar-baz {

  color: red;

}"
`;

exports[`Loader display-name 1: DisplayName.css 1`] = `
".display-name {

  color: ultra-red;

}"
`;

exports[`Loader display-name js : Compiled JS 1`] = `
"import styled from 'astroturf';

const Foo = () => {};
const bar = 'Bar';
Foo[bar]['baz'] = styled('div', null, \\"FooBarBaz\\", require(\\"./display-name-FooBarBaz.css\\"), \\"foo-bar-baz\\", \\"fooBarBaz\\");

export default styled(FancyBox, null, \\"DisplayName\\", require(\\"./DisplayName.css\\"), \\"display-name\\", \\"displayName\\");
"
`;

exports[`Loader element-shorthand 0: element-shorthand-FancyBox.css 1`] = `
".fancy-box {

  color: red;
  width: 75px;

  @media (min-width: 420px) {
    width: 96px;
    height: 96px;
  }

  &.primary {
    background: white;
    color: palevioletred;
  }

}"
`;

exports[`Loader element-shorthand js : Compiled JS 1`] = `
"import styled from 'astroturf';

const SIZE = 75;

const FancyBox = styled(\\"div\\", null, \\"FancyBox\\", require(\\"./element-shorthand-FancyBox.css\\"), \\"fancy-box\\", \\"fancyBox\\");
"
`;

exports[`Loader example 0: example-styles.less 1`] = `
"
  .foo {
    color: red;
    width: 75px;
  }

  .bar {
    composes: foo;

    font-weight: bold;
    font-style: italic;
  }
"
`;

exports[`Loader example js : Compiled JS 1`] = `
"

const SIZE = 75;

const styles = require('./example-styles.less');
"
`;

exports[`Loader hoist-import 0: hoist-import-ButtonBase.css 1`] = `
"@import '~./styles/mixins.scss';
.button-base {

  

  display: inline-flex;
  align-items: center;
  justify-content: center;
  border: 1px solid transparent;

}"
`;

exports[`Loader hoist-import js : Compiled JS 1`] = `
"import { styled } from 'astroturf';

const ButtonBase = styled('button', null, \\"ButtonBase\\", require(\\"./hoist-import-ButtonBase.css\\"), \\"button-base\\", \\"buttonBase\\");
"
`;

exports[`Loader imports-css 0: imports-css-styles.css 1`] = `
"
  .blue {
    color: blue;
  }
"
`;

exports[`Loader imports-css js : Compiled JS 1`] = `
"

const styles = require('./imports-css-styles.css');
"
`;

exports[`Loader imports-mixed 0: imports-mixed-styles.css 1`] = `
"
  .blue {
    color: blue;
  }
"
`;

exports[`Loader imports-mixed js : Compiled JS 1`] = `
"import styled from 'astroturf'

const styles = require('./imports-mixed-styles.css');
"
`;

exports[`Loader index 0: index-ButtonBase.css 1`] = `
"@import '~./styles/mixins.scss';
.button-base {

  

  display: inline-flex;
  align-items: center;
  justify-content: center;
  border: 1px solid transparent;

}"
`;

exports[`Loader index 1: Fixtures.css 1`] = `
".fixtures {

  color: ultra-red;

}"
`;

exports[`Loader index js : Compiled JS 1`] = `
"import styled from 'astroturf';

const ButtonBase = styled('button', null, \\"ButtonBase\\", require(\\"./index-ButtonBase.css\\"), \\"button-base\\", \\"buttonBase\\");

export default styled(FancyBox, null, \\"Fixtures\\", require(\\"./Fixtures.css\\"), \\"fixtures\\", \\"fixtures\\");
"
`;

exports[`Loader multiple-components 0: multiple-components-FancyBox.css 1`] = `
".fancy-box {

  color: red;
  width: 75px;

  @media (min-width: 420px) {
		width: 96px;
		height: 96px;
	}

  &.primary {
    background: white;
    color: palevioletred;
  }

}"
`;

exports[`Loader multiple-components 1: multiple-components-FancyHeader.css 1`] = `
".fancy-header {

  color: red;
  width: 75px;

  &.primary {
    background: white;
    color: palevioletred;
  }

}"
`;

exports[`Loader multiple-components js : Compiled JS 1`] = `
"import styled from 'astroturf'

const SIZE = 75;

// prettier-ignore
const FancyBox = styled('div', null, \\"FancyBox\\", require(\\"./multiple-components-FancyBox.css\\"), \\"fancy-box\\", \\"fancyBox\\");

const FancyHeader = styled('h2', null, \\"FancyHeader\\", require(\\"./multiple-components-FancyHeader.css\\"), \\"fancy-header\\", \\"fancyHeader\\");
"
`;

exports[`Loader no-global js : Compilation Error 1`] = `
"const SIZE = 75;

const styles = less\`
  .foo {
    color: red;
    width: \${SIZE}px;
  }

  .bar {
    composes: foo;

    font-weight: bold;
    font-style: italic;
  }
\`;
"
`;

exports[`Loader pass-options 0: pass-options-FancyBox.css 1`] = `
".fancy-box {

  color: red;
  width: 75px;

  @media (min-width: 420px) {
    width: 96px;
    height: 96px;
  }

  &.primary {
    background: white;
    color: palevioletred;
  }

}"
`;

exports[`Loader pass-options 1: pass-options-FancierBox.css 1`] = `
".fancier-box {

  color: ultra-red;

}"
`;

exports[`Loader pass-options js : Compiled JS 1`] = `
"import styled from 'astroturf';

const SIZE = 75;

// prettier-ignore
const FancyBox = styled('div', {
  someOption: true
}, \\"FancyBox\\", require(\\"./pass-options-FancyBox.css\\"), \\"fancy-box\\", \\"fancyBox\\");

const options = {};

const FancierBox = styled(FancyBox, options, \\"FancierBox\\", require(\\"./pass-options-FancierBox.css\\"), \\"fancier-box\\", \\"fancierBox\\");
"
`;

exports[`Loader styled 0: styled-FancyBox.css 1`] = `
".fancy-box {

  color: red;
  width: 75px;

  @media (min-width: 420px) {
    width: 96px;
    height: 96px;
  }

  &.primary {
    background: white;
    color: palevioletred;
  }

}"
`;

exports[`Loader styled 1: styled-FancierBox.css 1`] = `
".fancier-box {

  color: ultra-red;

}"
`;

exports[`Loader styled js : Compiled JS 1`] = `
"import styled from 'astroturf';

const SIZE = 75;

// prettier-ignore
const FancyBox = styled('div', null, \\"FancyBox\\", require(\\"./styled-FancyBox.css\\"), \\"fancy-box\\", \\"fancyBox\\");

const FancierBox = styled(FancyBox, null, \\"FancierBox\\", require(\\"./styled-FancierBox.css\\"), \\"fancier-box\\", \\"fancierBox\\");
"
`;

exports[`Loader styled-global 0: styled-global-FancyBox.css 1`] = `
".fancy-box {

  color: red;
  width: 75px;
  &.primary {
    background: white;
    color: palevioletred;
  }

}"
`;

exports[`Loader styled-global 1: styled-global-FancierBox.css 1`] = `
".fancier-box {

  color: ultra-red;

}"
`;

exports[`Loader styled-global js : Compiled JS 1`] = `
"// prettier-ignore
const FancyBox = styled('div', null, \\"FancyBox\\", require(\\"./styled-global-FancyBox.css\\"), \\"fancy-box\\", \\"fancyBox\\");

const FancierBox = styled(FancyBox, null, \\"FancierBox\\", require(\\"./styled-global-FancierBox.css\\"), \\"fancier-box\\", \\"fancierBox\\");
"
`;

exports[`Loader styled-no-assignment js : Compilation Error 1`] = `
"The output of this styled component is never used. Either assign it to a variable or export it.
[0m[31m[1m>[22m[39m[90m 1 | [39mstyled([32m'p'[39m)[32m\`[39m[0m
[0m [90m   | [39m[31m[1m^[22m[39m[0m
[0m [90m 2 | [39m[32m  color: blue;[39m[0m
[0m [90m 3 | [39m[32m\`[39m[33m;[39m[0m
[0m [90m 4 | [39m[0m"
`;

exports[`Loader styled-tag 0: styled-tag-FancyBox.css 1`] = `
".fancy-box {

  color: red;
  width: 75px;
  &.primary {
    background: white;
    color: palevioletred;
  }

}"
`;

exports[`Loader styled-tag 1: styled-tag-FancierBox.css 1`] = `
".fancier-box {

  color: ultra-red;

}"
`;

exports[`Loader styled-tag js : Compiled JS 1`] = `
"import s from 'astroturf';

// prettier-ignore
const FancyBox = styled('div', null, \\"FancyBox\\", require(\\"./styled-tag-FancyBox.css\\"), \\"fancy-box\\", \\"fancyBox\\");

const FancierBox = styled(FancyBox, null, \\"FancierBox\\", require(\\"./styled-tag-FancierBox.css\\"), \\"fancier-box\\", \\"fancierBox\\");
"
`;

exports[`Loader typescript 0: typescript-styles.css 1`] = `
"
  .foo {
    color: red;
    width: 75px;
  }

  .bar {
    composes: foo;

    font-weight: bold;
    font-style: italic;
  }
"
`;

exports[`Loader typescript js : Compiled JS 1`] = `
"
import React from 'react';

const SIZE = 75;

const styles = require('./typescript-styles.css');

interface PropsType {
  children: JSX.Element;
  name: string;
}

function fooGood<T extends { x: number }>(obj: T): T {
  console.log(Math.abs(obj.x));
  return obj;
}

class Component extends React.Component<PropsType, {}> {
  render() {
    return <h2>{this.props.children}</h2>;
  }
}

// OK
"
`;

exports[`Loader with-props 0: with-props-RedPasswordInput.css 1`] = `
".red-password-input {

  background-color: red;

}"
`;

exports[`Loader with-props js : Compiled JS 1`] = `
"import { styled } from 'astroturf';
import withProps from 'recompose/withProps';

const RedPasswordInput = withProps({
  type: 'password',
})(styled('input', null, \\"RedPasswordInput\\", require(\\"./with-props-RedPasswordInput.css\\"), \\"red-password-input\\", \\"redPasswordInput\\"));
"
`;
